File: haskell/021/Main.hs
{-
    ProjectEuler 012
    Haskell / Main.hs
    copyright (c) 2015 Susisu

    GHC 7.8.3
    `ghc Main.hs -o bin/main`
-}

module Main where

main :: IO ()
main = print $ itr 10000 2 0
    where
        itr n i s
            | i < n =
                let d = sumDivs i in
                if d /= i && sumDivs d == i then
                    itr n (i + 1) (s + i)
                else
                    itr n (i + 1) s
            | otherwise = s

sumDivs :: Int -> Int
sumDivs 1 = 0
sumDivs n = 1 + itr 2 0 + if r * r == n then r else 0
    where
        r = ceiling . sqrt $ fromIntegral n
        itr i s
            | i < r =
                if n `mod` i == 0 then
                    itr (i + 1) (s + i + n `div` i)
                else
                    itr (i + 1) s
            | otherwise = s

File: ocaml/021/main.ml
(*
    ProjectEuler 021
    OCaml / main.ml
    copyright (c) 2016 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main main.ml`
*)

let d n =
    if n <= 1 then 0
    else
        let rec itr s d =
            if n = d then s
            else if n mod d = 0 then itr (s + d) (d + 1)
            else itr s (d + 1)
        in
        itr 0 1

let main _ =
    let rec itr s n =
        if n >= 10000 then s
        else
            let m = d n in
            if n <> m && n = d m then itr (s + n) (n + 1)
            else itr s (n + 1)
    in
    let s = itr 0 1 in
    print_int s;
    print_newline ()

let _ = main ()
